### PR TITLE
Refactor RedBlackTree

### PR DESCRIPTION
**Pull request by ljw7975**
Refactor RedBlackTree

### CODE REVIEW

File: src/main/java/Tree/RedBlackTree.java [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+
+        if (parent != null && !parent.getColour()) {
+
+            if (uncle != null && !uncle.getColour()) {
+                parent.setColour(true);
+                uncle.setColour(true);
+                gp.setColour(false);
+                adjustAfterInsert(gp);
+            }
+            else if (parent.getKey() < gp.getKey()) {
```
Comment (ljj7975) [inline src/main/java/Tree/RedBlackTree.java:115]: parent is red but uncle is black (or missing?)

File: src/main/java/Tree/RedBlackTree.java [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+                    rightRotation(parent);
+                }
+                parent.setColour(true);
+                gp.setColour(false);
+                leftRotation(gp);
+            }
+        }
+        RedBlackNode<T> rt = (RedBlackNode<T>) root;
+        rt.setColour(true);
+        root = rt;
```
Comment (ljj7975) [inline src/main/java/Tree/RedBlackTree.java:134]: why not just root.setColor(true)?

Also, what if root changes due to rotation??

File: src/main/java/Tree/RedBlackTree.java [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+                        parent_node.setRightNode(null);
+                    }
+
+                    nodeToUpdate = deleteNode;
+                    // If deleting node is a leaf and black then mark it as double black
+                    if (nodeToUpdate.getColour()) {
+                        nodeToUpdate.setDoubleBlack(true);
+                    }
+                }
+            } else {
```
Comment (ljj7975) [inline src/main/java/Tree/RedBlackTree.java:167]: comment saying predecessor exists

File: src/main/java/Tree/RedBlackTree.java [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+                swapWithPredecessor(deleteNode, predecessor);
+                nodeToUpdate = (RedBlackNode<T>) predecessor.getParentNode();
+                if (predecessor.getColour() && predecessor.getRightNode() == null && predecessor.getLeftNode() == null) {
+                    handleDoubleBlack(predecessor);
+                }
+            }
+        } else {
```
Comment (ljj7975) [inline src/main/java/Tree/RedBlackTree.java:174]: same

File: src/main/java/Tree/RedBlackTree.java [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
+    public void adjustAfterDelete(RedBlackNode<T> node) {
+        if (node.isDoubleBlack()) {
+            handleDoubleBlack(node);
+        } else if (node.getColour()) {
+            deleteCase1(node);
+        }
+    }
+
+    public void handleDoubleBlack(RedBlackNode<T> node) {
```
Comment (ljj7975) [inline src/main/java/Tree/RedBlackTree.java:195]: For some of the functions that is hard to understand the context,
please leave comments on what they are supposed to do.

File: src/main/java/Tree/RedBlackTree.java [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+                }
+            } else if (node.getKey() < parent.getKey()) {
+                if (siblingRight != null && !siblingRight.getColour()) {
+                    leftRotation(parent);
+                } else if (siblingLeft != null) {
+                    rightRotation(sibling);
+                    leftRotation(parent);
+                    parent.setColour(true);
+                }
+            } else {
```
Comment (ljj7975) [inline src/main/java/Tree/RedBlackTree.java:218]: when would you fall into this case?

File: src/main/java/Tree/RedBlackTree.java [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+                if (parent.getLeftNode() == null) {
+                    leftRotation(parent);
+                } else if (parent.getRightNode() == null) {
+                    rightRotation(parent);
+                }
+            }
+        } else if (!sibling.getColour()) {
```
Comment (ljj7975) [inline src/main/java/Tree/RedBlackTree.java:225]: do I need this else if? why not just else with comment?

File: src/main/java/Tree/RedBlackTree.java [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+            if (node.getKey() < parent.getKey()) {
+                leftRotation(parent);
+            } else {
+                rightRotation(parent);
+            }
+            parent.setColour(false);
+            sibling.setColour(true);
+            handleDoubleBlack(node);
```
Comment (ljj7975) [inline src/main/java/Tree/RedBlackTree.java:233]: we call handleDoubleBlack on the same node twice?

File: src/main/java/Tree/RedBlackTree.java [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+        }
+    }
+
+    public void deleteCase1(RedBlackNode<T> node) {
```
Comment (ljj7975) [inline src/main/java/Tree/RedBlackTree.java:237]: for all deleteCase#, give more descriptive names or add comments

Comment (ljj7975) [reply src/main/java/Tree/RedBlackTree.java:237]: why can't this be combined with `adjustAfterDelete` and `deleteCase3`?
this is just one single deleteCase

File: src/main/java/Tree/RedBlackTree.java [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+        } else if (!parent.getColour() && sibling.getColour()) {
+            if (siblingLeft != null && siblingRight != null && siblingLeft.getColour() && siblingRight.getColour()) {
+                if (!node.getColour()) {
+                    sibling.setColour(false);
+                    parent.setColour(true);
+                }
+            }
+        } else {
+            deleteCase4(node);
+        }
```
Comment (ljj7975) [inline src/main/java/Tree/RedBlackTree.java:278]: please add description for each if cases

File: src/main/java/Tree/RedBlackTree.java (hunk 2 above)
Comment (ljw7975) [reply src/main/java/Tree/RedBlackTree.java:134]: left/right rotation handle root changing cases

File: src/main/java/Tree/RedBlackTree.java [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+        RedBlackNode<T> sibling = null;
+        RedBlackNode<T> parent = (RedBlackNode<T>) node.getParentNode();
+        assert (node == null || parent == null) : "node and parent must not be null";
+
+        if (node.getKey() < parent.getKey()) {
+            sibling = (RedBlackNode<T>) parent.getRightNode();
+        } else if (node.getKey() > parent.getKey()){
+            sibling = (RedBlackNode<T>) parent.getLeftNode();
+        }
+        else {
```
[the remark this answers is not in the document]
Comment (ljw7975) [reply src/main/java/Tree/RedBlackTree.java:47]: when handling doubleBlack cases

File: src/main/java/Tree/RedBlackTree.java (hunk 2 above)
Comment (ljw7975) [reply src/main/java/Tree/RedBlackTree.java:134]: cannot cast setColour(RedBlackNode function) on BST node root.

File: src/main/java/Tree/RedBlackTree.java (hunk 8 above)
Comment (ljw7975) [reply src/main/java/Tree/RedBlackTree.java:233]: yes, when double black has red sibling, rotate first to make sibling black and then perform double black handling.

**Pull request merged**